File: examples/js/nft-generate-mint.js
/**
 * ChainGPT AI NFT Generator — Generate + Mint Example
 *
 * Demonstrates: image generation, prompt enhancement, NFT minting on BSC
 * Install: npm install @chaingpt/nft dotenv
 */
import 'dotenv/config';
import { Nft, Errors } from '@chaingpt/nft';
import fs from 'fs';

const nft = new Nft({ apiKey: process.env.CHAINGPT_API_KEY });

// 1. Generate a single image
async function generateImage() {
  const result = await nft.generateImage({
    prompt: 'A cyberpunk samurai standing on a neon-lit blockchain bridge',
    model: 'nebula_forge_xl',
    height: 1024,
    width: 1024,
    steps: 25,
    enhance: '1x',
    style: 'neon-punk'
  });
  // Save the image bytes
  const buffer = Buffer.from(result.data);
  fs.writeFileSync('generated-nft.png', buffer);
  console.log('Image saved to generated-nft.png');
}

// 2. Full pipeline: Enhance → Generate → Mint
async function generateAndMint() {
  // Step 1: Enhance the prompt
  const enhanced = await nft.enhancePrompt({
    prompt: 'a dragon in space'
  });
  console.log('Enhanced prompt:', enhanced.data.enhancedPrompt);

  // Step 2: Generate NFT (synchronous — waits for completion)
  const genResult = await nft.generateNft({
    prompt: enhanced.data.enhancedPrompt,
    model: 'velogen',
    height: 512,
    width: 512,
    steps: 3,
    enhance: '1x',
    walletAddress: process.env.WALLET_ADDRESS,
    chainId: 56,  // BSC Mainnet
    amount: 1
  });
  console.log('Collection ID:', genResult.data.collectionId);

  // Step 3: Mint the NFT
  const mintResult = await nft.mintNft({
    collectionId: genResult.data.collectionId,
    name: 'Space Dragon #1',
    description: 'An AI-generated cosmic dragon NFT',
    symbol: 'DRGN',
    ids: [1]
  });
  console.log('Mint result:', mintResult);
  console.log('IPFS Image:', mintResult.image);
}

// 3. Async queue-based generation (for large batches)
async function batchGenerate() {
  const result = await nft.generateNftWithQueue({
    prompt: 'Abstract geometric crystal formations',
    model: 'nebula_forge_xl',
    height: 1024,
    width: 1024,
    steps: 30,
    walletAddress: process.env.WALLET_ADDRESS,
    chainId: 137,  // Polygon
    amount: 5
  });
  const collectionId = result.data.collectionId;
  console.log('Queued:', collectionId);

  // Poll progress
  let done = false;
  while (!done) {
    const progress = await nft.getNftProgress({ collectionId });
    console.log(`Progress: ${progress.progress || 0}%`);
    if (progress.data?.generated) {
      done = true;
      console.log('Images:', progress.data.images);
    } else {
      await new Promise(r => setTimeout(r, 3000));
    }
  }
}

// 4. List supported chains
async function listChains() {
  const chains = await nft.getChains(false); // mainnet
  console.log('Supported chains:', chains);
}

(async () => {
  try {
    await generateImage();
    // await generateAndMint();  // Uncomment when WALLET_ADDRESS is set
    // await batchGenerate();
    await listChains();
  } catch (error) {
    if (error instanceof Errors.NftError) {
      console.error('NFT Error:', error.message);
    } else {
      throw error;
    }
  }
})();

File: examples/python/nft_generate_mint.py
"""
ChainGPT AI NFT Generator — Generate + Mint Example (Python)

Demonstrates: image generation, prompt enhancement, minting
Install: pip install chaingpt python-dotenv

NOTE: The Python SDK's actual class names, method signatures, and import
paths may differ from what is shown here. These examples illustrate the
*concepts* (image generation, minting, prompt enhancement) but are not
guaranteed to match the latest SDK release. Always check the official
ChainGPT documentation at https://docs.chaingpt.org for up-to-date
Python SDK usage.
"""
import asyncio
import os
import time
from dotenv import load_dotenv
from chaingpt.client import ChainGPTClient
from chaingpt.models.nft import (
    NFTGenerateImageRequestModel,
    NFTGenerateQueueRequestModel,
    NFTMintRequestModel,
    NFTEnhancePromptRequestModel
)
from chaingpt.types import NFTImageModel, ImageEnhanceOption

load_dotenv()
API_KEY = os.environ["CHAINGPT_API_KEY"]
WALLET = os.environ.get("WALLET_ADDRESS", "0x0000000000000000000000000000000000000000")


async def generate_image():
    """Generate a single AI image."""
    async with ChainGPTClient(api_key=API_KEY) as client:
        result = await client.nft.generate_image(NFTGenerateImageRequestModel(
            prompt="A neon samurai guarding a blockchain vault",
            model=NFTImageModel.NEBULA_FORGE_XL,
            height=1024,
            width=1024,
            steps=25,
            enhance=ImageEnhanceOption.ENHANCE_1X
        ))
        # Save image
        with open("generated_nft.png", "wb") as f:
            f.write(bytes(result.data))
        print("Image saved to generated_nft.png")


async def enhance_and_generate():
    """Enhance prompt then generate."""
    async with ChainGPTClient(api_key=API_KEY) as client:
        # Enhance
        enhanced = await client.nft.enhance_prompt(
            NFTEnhancePromptRequestModel(prompt="a dragon in space")
        )
        print(f"Enhanced: {enhanced.data.enhancedPrompt}")

        # Generate with queue
        gen = await client.nft.generate_nft_queue(NFTGenerateQueueRequestModel(
            prompt=enhanced.data.enhancedPrompt,
            model=NFTImageModel.VELOGEN,
            height=512, width=512, steps=3,
            walletAddress=WALLET,
            chainId=56,  # BSC
            amount=1
        ))
        collection_id = gen.collectionId
        print(f"Queued: {collection_id}")

        # Poll progress
        while True:
            progress = await client.nft.get_progress(collection_id)
            if hasattr(progress, 'data') and progress.data and progress.data.generated:
                print(f"Done! Images: {progress.data.images}")
                break
            print("Generating...")
            await asyncio.sleep(3)

        # Mint
        mint = await client.nft.mint_nft(NFTMintRequestModel(
            collectionId=collection_id,
            name="Space Dragon #1",
            description="AI-generated cosmic dragon",
            symbol="DRGN",
            ids=[1]
        ))
        print(f"Minted! IPFS: {mint.image}")


async def list_chains():
    """List supported blockchains."""
    async with ChainGPTClient(api_key=API_KEY) as client:
        chains = await client.nft.get_chains(test_net=False)
        print("Supported chains:", chains)


async def main():
    await generate_image()
    # await enhance_and_generate()  # Uncomment with real WALLET_ADDRESS
    await list_chains()

if __name__ == "__main__":
    asyncio.run(main())
